wasm/validator/src/recursive.rs

```diff
@@ -0,0 +1,290 @@
+/// with `validate_query` function;
+/// 3. Validate simple instruction (aka [`DirectInstruction`](DirectInstruction)) with
+/// `validate_instruction` function;
+///
+/// # Query validation
+///
+/// Query validation is done directly with `validate_query` function.
+///
+/// # Example
+///
+/// ```
+///#![no_std]
+/// #![no_main]
+///
+/// #[cfg(not(test))]
+/// extern crate panic_halt;
+///
+/// use iroha_validator::prelude::*;
+///
+/// #[entrypoint(params = "[authority, operation]")]
+/// pub fn validate(
+///     authority: <Account as Identifiable>::Id,
+///     operation: NeedsPermissionBox,
+/// ) -> Verdict {
+///     validate_recursively(
+///         authority,
+///         operation,
+///         |_authority, _transaction| pass!(),
+///         validate_instruction,
+///         |_authority, _query| pass!(),
+///     )
+/// }
+///
+/// fn validate_instruction(
+///     authority: &<Account as Identifiable>::Id,
+///     instruction: DirectInstruction,
+/// ) -> Verdict {
+///     todo!("Your instruction validation logic here")
+/// }
+/// ```
+pub fn validate_recursively<T, I, Q>(
+    authority: <Account as Identifiable>::Id,
+    operation: NeedsPermissionBox,
+    validate_transaction: T,
+    validate_instruction: I,
+    validate_query: Q,
+) -> Verdict
+where
+    T: FnOnce(&<Account as Identifiable>::Id, TransactionInfo) -> Verdict,
+    I: Fn(&<Account as Identifiable>::Id, DirectInstruction) -> Verdict + Copy,
+    Q: Fn(&<Account as Identifiable>::Id, QueryBox) -> Verdict + Copy,
+{
+    match operation {
+        NeedsPermissionBox::Transaction(transaction) => {
+            let (transaction_info, instructions) = TransactionInfo::disassemble_tx(transaction);
+            validate_transaction(&authority, transaction_info).and_then(|| {
+                let Executable::Instructions(instructions) = instructions else {
+                    pass!();
+                };
+
+                for isi in instructions {
+                    let verdict = validate_instruction_recursively(
+                        &authority,
+                        isi.clone(),
+                        validate_instruction,
+                        validate_query,
+                    );
+                    if verdict.is_deny() {
+                        return verdict;
+                    }
+                    isi.execute()
+                }
+                pass!();
+            })
+        }
+        NeedsPermissionBox::Instruction(instruction) => {
+            let verdict = validate_instruction_recursively(
+                &authority,
+                instruction.clone(),
+                validate_instruction,
+                validate_query,
+            );
+            if verdict.is_pass() {
+                instruction.execute()
+            }
+            verdict
+        }
+        NeedsPermissionBox::Query(query) => validate_query(&authority, query),
+    }
+}
```

Comment:
following the discussion [here](https://github.com/hyperledger/iroha/pull/3329/files#r1155295952). This is how I would do it(roughly):

```suggestion
pub fn validate_recursively<T, I, Q>(
    authority: <Account as Identifiable>::Id,
    operation: NeedsPermissionBox,
    validate_transaction: T,
    validate_instruction: I,
    validate_query: Q,
) -> Verdict
where
    T: FnOnce(&<Account as Identifiable>::Id, Transaction) -> Verdict,
    I: Fn(&<Account as Identifiable>::Id, DirectInstruction) -> Verdict + Copy,
    Q: Fn(&<Account as Identifiable>::Id, QueryBox) -> Verdict + Copy,
{
    match operation {
        NeedsPermissionBox::Transaction(transaction) => {
            validate_transaction(&authority, transaction)
        }
        NeedsPermissionBox::Instruction(instruction) => {
            let verdict = validate_instruction_recursively(
                &authority,
                instruction.clone(),
                validate_instruction,
                validate_query,
            );
            if verdict.is_pass() {
                instruction.execute()
            }
            verdict
        }
        NeedsPermissionBox::Query(query) => validate_query(&authority, query),
    }
}

fn validate_transaction(authority: &AccountId, transaction: Transaction) -> Verdict
    let Executable::Instructions(instructions) = instructions else {
        pass!();
    };

    for isi in instructions {
        let verdict = validate_instruction_recursively(
            &authority,
            isi.clone(),
            validate_instruction,
            validate_query,
        );
        if verdict.is_deny() {
            return verdict;
        }
        isi.execute()
    }

    pass!();
}
```

Reply:
I don't get it. You have two `validate_transaction()`. One as a function parameter and another as a function below. So it seems like one of them is not used

Reply:
everything is ok. The one that is a function is what you will give to the `validate_recursively` as a parameter

Reply:
So how do you think users should validate transaction then? They should write recurscive unwrapping every time?

Reply:
and if the user wants they can provide a different function, but this one is a good default for them

Reply:
Done